File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;
class Solution {
    private boolean isCorrect(char[] str, int offset){
        Stack<Character> stk = new Stack<>();
        
        for(int i = 0; i < str.length; i++){
            char c = str[(i + offset) % str.length];
            switch(c){
                    case '(' -> stk.push(')');
                    case '{' -> stk.push('}');
                    case '[' -> stk.push(']');
                    case ')', '}', ']' ->{
                        if(stk.isEmpty()) return false;
                        if(stk.pop() != c) return false;
                    }
            }
        }
        
        return stk.isEmpty();
    }
    public int solution(String s) {
        char[] str = s.toCharArray();
        
        int ret = 0;
        for(int offset = 0; offset < str.length; offset++){
            if(isCorrect(str, offset)) ret++;
        }
        return ret;
    }
}
/*
1. 회전
2. 짝짓기

*/

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def check(s):
    stk = []
    for c in s:
        if not stk and (c == ')' or c == '}' or c == ']'):
             return False
        if c == '(' or c == '[' or c == '{':
            stk.append(c)
        elif c == ')' and stk[len(stk)-1] == '(':
            stk.pop()
        elif c == ']' and stk[len(stk)-1] == '[':
            stk.pop()
        elif c == '}' and stk[len(stk)-1] == '{':
            stk.pop()
    return False if stk else True
def solution(s):
    ret = 0
    for i in range(len(s)):
        if(check(s[i:]+s[:i])): ret += 1
    return ret